Reduce WTF_ALLOW_UNSAFE_BUFFER_USAGE usage in WebCore/platform/graphics/cg

cdumez opened the pull request:
#### d1f65a28322f79ee84fbd436a3e0c331587ba358
<pre>
Reduce WTF_ALLOW_UNSAFE_BUFFER_USAGE usage in WebCore/platform/graphics/cg
<a href="https://bugs.webkit.org/show_bug.cgi?id=285725">https://bugs.webkit.org/show_bug.cgi?id=285725</a>

Reviewed by Darin Adler.

* Source/WebCore/platform/graphics/cg/ColorCG.cpp:
(WebCore::componentsSpan):
(WebCore::roundAndClampToSRGBALossy):
(WebCore::Color::createAndPreserveColorSpace):
* Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::setCGStyle):
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.cpp:
(WebCore::ImageBufferCGBitmapBackend::create):
* Source/WebCore/platform/graphics/cg/ImageBufferUtilitiesCG.cpp:
(WebCore::encode):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::ImageDecoderCG::decodeUTI):
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::transcodeImage):
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::pointsSpan):
(WebCore::addUnevenCornersRoundedRect):
(WebCore::pathElementApplierCallback):
(WebCore::copyClosingSubpathsApplierFunction):
* Source/WebCore/platform/graphics/cg/ShareableBitmapCG.mm:
(WebCore::ShareableBitmap::createFromImagePixels):

Canonical link: <a href="https://commits.webkit.org/288739@main">https://commits.webkit.org/288739@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7160688d2e94ec37d022768d283de299844af3e8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84266 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3887 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38570 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89342 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35275 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86351 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3975 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11862 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65544 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23375 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87312 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2980 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76572 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45835 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2930 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30802 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34324 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73833 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31571 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90724 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11533 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/8374 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73987 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11759 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72397 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73189 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17512 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15959 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2898 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13041 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11485 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16961 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11334 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14810 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13107 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.cpp:
```diff
@@ -54,28 +53,31 @@ std::unique_ptr<ImageBufferCGBitmapBackend> ImageBufferCGBitmapBackend::create(c
         return nullptr;
 
-    void* data;
     unsigned bytesPerRow = 4 * backendSize.width();
```
Why start checking only _after_ multiplying by 4? We can start using `CheckedSize` a little sooner.

cdumez replied:
Indeed, I missed that. I'll fix before landing.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.cpp:
```diff
@@ -54,28 +54,34 @@ std::unique_ptr<ImageBufferCGBitmapBackend> ImageBufferCGBitmapBackend::create(c
 
-    if (!tryFastCalloc(backendSize.height(), bytesPerRow).getValue(data))
+    CheckedSize numBytes = backendSize.height();
+    numBytes *= bytesPerRow;
```
Not sure why the case above uses `checkedProduct` and this case uses `*=` but both seem OK

cdumez replied:
Ok, I'll make it consistent.

darinadler commented on Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
```diff
@@ -632,9 +630,9 @@ String ImageDecoderCG::decodeUTI(CGImageSourceRef imageSource, const SharedBuffe
+        auto value = reinterpretCastSpanStartTo<const unsigned>(span.subspan(index * sizeof(unsigned)));
         // Numbers in the file are BigEndian.
-        return flipBytesIfLittleEndian(boxBytes[index], isLittleEndian);
+        return flipBytesIfLittleEndian(value, isLittleEndian);
```
Not new to this patch: This a strange way to write this, the way it involves a named boolean argument. I wish we used a more straightforward idiom.

For example, if I was writing this I would read 4 bytes out as a big-endian value, which we must do in at least some places in WebKit. That would be something like `readBigEndian<uint32_t>` function that takes a `std::span<uint8_t>` argument, avoiding both the `reinterpret_cast` and the "flip bytes" concept.

But if we are going to do it like this, with "flipping", the `flipBytesIfLittleEndian` function could have an overload that doesn’t take a boolean argument. Given the function name it seems logical that it would flip the bytes in the passed-in value iff the platform’s byte ordering is little-endian.

darinadler commented on Source/WebCore/platform/graphics/cg/ShareableBitmapCG.mm:
```diff
@@ -129,28 +128,26 @@
-    const auto* bytes = byteCast<uint8_t>(CFDataGetBytePtr(pixels.get()));
-    CheckedUint32 sizeInBytes = CFDataGetLength(pixels.get());
-    if (!bytes || !sizeInBytes || sizeInBytes.hasOverflowed())
+    auto bytes = WTF::span(pixels.get());
+    if (bytes.empty() || CheckedUint32(bytes.size()).hasOverflowed())
         return nullptr;
```
Do you understand where the "must not overflow 32 bits" requirement comes from?

cdumez replied:
I do not, I'll try and figure it out. ShareableBitmapConfiguration seems to be consistently using uint32_t (or checked versions) everywhere.

cdumez committed:
Reduce WTF_ALLOW_UNSAFE_BUFFER_USAGE usage in WebCore/platform/graphics/cg
https://bugs.webkit.org/show_bug.cgi?id=285725

Reviewed by Darin Adler.

* Source/WebCore/platform/graphics/cg/ColorCG.cpp:
(WebCore::componentsSpan):
(WebCore::roundAndClampToSRGBALossy):
(WebCore::Color::createAndPreserveColorSpace):
* Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::setCGStyle):
* Source/WebCore/platform/graphics/cg/ImageBufferCGBitmapBackend.cpp:
(WebCore::ImageBufferCGBitmapBackend::create):
* Source/WebCore/platform/graphics/cg/ImageBufferUtilitiesCG.cpp:
(WebCore::encode):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::ImageDecoderCG::decodeUTI):
* Source/WebCore/platform/graphics/cg/ImageUtilitiesCG.cpp:
(WebCore::transcodeImage):
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
* Source/WebCore/platform/graphics/cg/PathCG.cpp:
(WebCore::pointsSpan):
(WebCore::addUnevenCornersRoundedRect):
(WebCore::pathElementApplierCallback):
(WebCore::copyClosingSubpathsApplierFunction):
* Source/WebCore/platform/graphics/cg/ShareableBitmapCG.mm:
(WebCore::ShareableBitmap::createFromImagePixels):

Canonical link: https://commits.webkit.org/288739@main
